Add policy support for named ports

jrajahalme opened the pull request:
Add support for container port names in policy ports. With this change, in addition to using a L4 port number (like "80") in the network policy (K8s Network Policy or Cilium Network Policy), the port number can be expressed as a name specified in any one of the POD specs in the cluster. If the port name is not specified in any container spec, the traffic specified in the network policy rule can't be allowed, but will be allowed as soon as a container with a matching port name is added.

Port names are stored from a POD spec on the node in which the POD is deployed, from where then name/port mappings are distributed to other nodes via CEP CRDs.

```release-note
Support for named k8s container ports is added to both K8s Network Policies and Cilium Network Policies.
```

aanm committed:
pkg/k8s: rename HostPorts to ContainerPorts

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/k8s: copy init containers

Init Containers should also be part of our Pod spec since the network
namespace is shared for those containers as well otherwise we won't
be able to enforce hostport or any other features in init containers

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme committed:
policy: Pass L4Filter to AccumulateMapChanges()

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
k8s: Add named ports to CEP

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Support named ports in policy trace

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -1157,13 +1157,16 @@ var (
 				Type: "string",
-				// uint16 string regex
+				// uint16 or port name string regex
 				Pattern: `^(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|` +
-					`[1-5][0-9]{4}|[0-9]{1,4})$`,
+					`[1-5][0-9]{4}|[0-9]{1,4}|` +
+					`([a-zA-Z0-9]-?)*[a-zA-Z](-?[a-zA-Z0-9])*)$`,
```
This requires a bump in the version set in `CustomResourceDefinitionSchemaVersion`

jrajahalme replied:
Will do.

jrajahalme replied:
Both places? (`pkg/k8s/apis/cilium.io/v2/client/register.go` and `pkg/k8s/apis/cilium.io/v2/register.go`)?

jrajahalme replied:
Actually, it looks like one in `pkg/k8s/apis/cilium.io/v2/register.go` is not used for anything anymore.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -1057,6 +1064,42 @@ func (e *Endpoint) SetK8sNamespace(name string) {
+			continue // skip unnamed ports
+		}
+		if !api.IsNamedPort(cp.Name) {
+			log.WithField(logfields.PortName, cp.Name).Warning("ContainerPort: Invalid port name, not using as a named port")
```
How can the user visualize this error better? Thinking on this better, this should never happened has the validation is done in k8s side.

jrajahalme replied:
Right, this is likely not needed, but still wanted to have logging if this happens.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -1057,6 +1064,42 @@ func (e *Endpoint) SetK8sNamespace(name string) {
+			log.WithField(logfields.PortName, cp.Name).Warning("ContainerPort: Invalid port name, not using as a named port")
+			continue
+		}
+		name := strings.ToLower(cp.Name)
```
I assume named ports are case insensitive?

jrajahalme replied:
Yes. They are defined as `IANA_SVC_NAME`, which is defined as case insensitive in 
https://tools.ietf.org/rfc/rfc6335.txt

aanm commented on pkg/policy/l4.go:
```diff
@@ -464,13 +500,21 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
+	portName := ""
+	p := uint64(0)
+	if api.IsNamedPort(port.Port) {
+		portName = port.Port
```
Does this mean that a named port can't contain only numbers?

jrajahalme replied:
Yes: 
```markdown
[rfc6335](https://tools.ietf.org/rfc/rfc6335.txt) port name (IANA_SVC_NAME)
: An alphanumeric (a-z, and 0-9) string, with a maximum length of 15 characters, with the '-' character allowed anywhere except the first or the last character or adjacent to another '-' character, it must contain at least a (a-z) character
```

aanm commented on pkg/u8proto/u8proto.go:
```diff
@@ -41,6 +41,7 @@ var protoNames = map[U8proto]string{
 }
 
 var ProtoIDs = map[string]U8proto{
+	"":       0,
```
Why this change?

jrajahalme replied:
Not sure any more... To map a missing proto name to ANY, but then in K8s ContainerPort's Protocol defaults to "TCP". I'll check if this is still needed & verify the default behavior.

jrajahalme replied:
It seems my assumption was that ContainerPort could have ANY protocol, but that is not the case, even though this PR would allow it. Our policy rules allow ANY protocol, which we internally treat as both UDP and TCP and generate two rules (one for each). One of them gets pruned away when the port name is resolved.

It might be better if I make this implementation less permissive and only allow one protocol, missing defaulting to TCP?

jrajahalme replied:
Changed empty proto string to default to TCP, but kept the internal processing that allows "ANY".

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -1583,6 +1586,9 @@ definitions:
       cluster-mesh:
         description: Status of ClusterMesh
         "$ref": "#/definitions/ClusterMeshStatus"
+      namedPorts:
```
What is the intent behind adding it to `cilium status`?

jrajahalme replied:
Half-baked idea of showing the existence of named ports there. Will remove.

jrajahalme replied:
Removed.

aanm commented on pkg/ipcache/ipcache.go:
```diff
@@ -54,6 +55,8 @@ type K8sMetadata struct {
 	Namespace string
 	// PodName is the Kubernetes pod name behind the IP
 	PodName string
+	// NamedPorts is the set of named ports for the pod
+	NamedPorts policy.NamedPortsMap
```
I would be extremely careful with this. Adding this field removes the ability to compare 2 variables created from this structure type with the `==` operator.

jrajahalme replied:
It has an `Equal` method already, I did expand it. Could also consider making `NamedPorts` a separate member/argument, although port names are semantically "k8s metadata".

aanm commented on pkg/ipcache/ipcache.go:
```diff
@@ -267,10 +318,51 @@ func (ipc *IPCache) Upsert(ip string, hostIP net.IP, hostKey uint8, k8sMeta *K8s
-		ipc.ipToK8sMetadata[ip] = *k8sMeta
+	if !metaEqual {
+		// Keep old named ports if new identity is from kv-store and has no named ports.
+		// This needs to be done to preserve the named ports as kv-store does not have them.
```
Can't we add it to the kvstore?

jrajahalme replied:
You tell me. I recall facing a comment that discouraged me from adding it.

jrajahalme replied:
```
// IPIdentityPair is a pairing of an IP and the security identity to which that
// IP corresponds. May include an optional Mask which, if present, denotes that
// the IP represents a CIDR with the specified Mask.
//
// WARNING - STABLE API
// This structure is written as JSON to the key-value store. Do NOT modify this
// structure in ways which are not JSON forward compatible.
type IPIdentityPair struct {
	IP           net.IP          `json:"IP"`
	Mask         net.IPMask      `json:"Mask"`
	HostIP       net.IP          `json:"HostIP"`
	ID           NumericIdentity `json:"ID"`
	Key          uint8           `json:"Key"`
	Metadata     string          `json:"Metadata"`
	K8sNamespace string          `json:"K8sNamespace,omitempty"`
	K8sPodName   string          `json:"K8sPodName,omitempty"`
}
```

jrajahalme replied:
Maybe I did not spend enough time figuring out what it takes to be "JSON forward compatible"?

tgraf replied:
You can add fields but you can't rename or remove existing ones.

jrajahalme replied:
So the consensus is to add the named ports to kv-store as well?

jrajahalme replied:
Done.

aanm reviewed (changes requested):
I didn't finish my review but I'm leaving some comments for now

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -145,11 +145,11 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
 // Must be called with endpoint.Mutex held.
 func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
-	if option.Config.DryMode {
+	if option.Config.DryMode || e.isProxyDisabled() {
```
Is this change related to named ports?

jrajahalme replied:
Not strictly, but instead of refactoring `updateProxyRedirect()` which was only called from here, I inlined it here instead. `e.isProxyDisabled()` came from there.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -1057,6 +1064,42 @@ func (e *Endpoint) SetK8sNamespace(name string) {
+			log.WithField(logfields.Port, cp.ContainerPort).Warning("ContainerPort: Port number out of 16-bit range")
+			continue
+		}
+		k8sPorts[name] = policy.NamedPort{
```
Should we warn on duplicates?

jrajahalme replied:
This is for the Endpoint, so these come from the POD spec, and I'd expect K8s to not allow duplicate names in the POD spec?

tgraf replied:
Not sure. I would warn so we at least know.

christarazi commented on daemon/cmd/endpoint.go:
```diff
@@ -151,24 +152,24 @@ func NewPutEndpointIDHandler(d *Daemon) PutEndpointIDHandler {
 // fetchK8sLabelsAndAnnotations wraps the k8s package to fetch and provide
 // endpoint metadata. It implements endpoint.MetadataResolverCB.
-func (d *Daemon) fetchK8sLabelsAndAnnotations(nsName, podName string) (labels.Labels, labels.Labels, map[string]string, error) {
+func (d *Daemon) fetchK8sLabelsAndAnnotations(nsName, podName string) ([]types.ContainerPort, labels.Labels, labels.Labels, map[string]string, error) {
```
Suggestion: could we consider making the return value of this function a struct?

jrajahalme replied:
I'll leave the judgement here to @aanm

aanm replied:
Yeah this needs to be addressed as a follow up of this PR. I have another PR that adds one more argument to this function. It's growing...

tgraf commented on pkg/policy/l4.go:
```diff
@@ -240,6 +242,31 @@ func (l4 *L4Filter) GetPort() uint16 {
+	if np.Proto != 0 && proto != np.Proto {
+		return 0, fmt.Errorf("incompatible proto")
+	}
+	if np.Port == 0 {
```
When can this happen? Shouldn't this be rejected during validation earlier?

jrajahalme replied:
Should not happen, but wanted to be safe as 0 port in policy logic indicates wildcarded L4 and it would be bad if 0 slipped through here.

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -179,15 +179,25 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
 				var revertFunc revert.RevertFunc
-				redirectPort, err, finalizeFunc, revertFunc = e.updateProxyRedirect(l4, proxyWaitGroup)
+
+				proxyID := ""
```
Nit: is there a specific reason this is initialized here?

jrajahalme replied:
It is equivalent to `var proxyID string`. It looked like a declaration is required due to `err` already being declared, but this can be simplified as:
```
@@ -172,7 +172,6 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
        for _, l4 := range m {
                if l4.IsRedirect() {
                        var redirectPort uint16
-                       var err error
                        // Only create a redirect if the proxy is NOT running in a sidecar
                        // container. If running in a sidecar container, just allow traffic
                        // to the port at L4 by setting the proxy port to 0.
@@ -180,8 +179,7 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
                                var finalizeFunc revert.FinalizeFunc
                                var revertFunc revert.RevertFunc
 
-                               proxyID := ""
-                               proxyID, err = e.ProxyID(e.desiredPolicy.NamedPortsMap, l4)
+                               proxyID, err := e.ProxyID(e.desiredPolicy.NamedPortsMap, l4)
                                if err != nil {
                                        // skip redirects for which a proxyID cannot be created.
                                        // Right now this happens only due to named port mapping not existing and
```

Thanks!

aanm replied:
Not addressed?

jrajahalme committed:
envoy: Fold rules for overlapping ports together

When using named ports that map to an already existing port number in
the policy we need to fold the rules together as otherwise Envoy will
NACK them due to the duplicate port number.

We probably should also detect mismatching L7 parsers here?

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
ipcache: Store named ports in kv-store

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on daemon/cmd/ipcache.go:
```diff
@@ -86,6 +86,7 @@ func (ipc *ipCacheDumpListener) OnIPIdentityCacheChange(modType ipcache.CacheMod
 			Source:    string(source.Kubernetes),
 			Namespace: k8sMeta.Namespace,
 			Name:      k8sMeta.PodName,
+			// TODO: Add Named ports??
```
Has this TODO been taken care of?

jrajahalme replied:
No, it is still an open question if we want the named ports in `IPListEntryMetadata`.

aanm replied:
Please add more information about it in this TODO (your username for example)

christarazi reviewed (commented):
LGTM besides a few minor comments

jrajahalme committed:
CI: Use named port "http-80" in k8s policy tests.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on cilium/cmd/policy_trace.go:
```diff
@@ -272,8 +273,11 @@ func getSecIDFromK8s(podName string) (string, error) {
 }
 
+// IsNamedPort is copied from pkg/policy/api/l4.go to not need to import the whole package.
+var IsNamedPort = regexp.MustCompile(`^([a-zA-Z0-9]-?)*[a-zA-Z](-?[a-zA-Z0-9])*$`).MatchString
```
Can't the regex of `pkg/policy/api/l4.go` live in a common package that is shared with this one?

jrajahalme replied:
Added new `pkg/iana` for this.

aanm commented on test/kubernetes-test.sh:
```diff
@@ -73,4 +73,4 @@ export KUBE_MASTER=192.168.36.11
 export KUBE_MASTER_URL="https://192.168.36.11:6443"
 
-${HOME}/go/bin/kubetest --test --test_args="--ginkgo.focus=NetworkPolicy --e2e-verify-service-account=false --host ${KUBE_MASTER_URL} --ginkgo.skip=(should.allow.egress.access.to.server.in.CIDR.block)|(should.allow.ingress.access.from.updated.pod)|(named.port)"
+${HOME}/go/bin/kubetest --test --test_args="--ginkgo.focus=NetworkPolicy.*on.one.named.port --e2e-verify-service-account=false --host ${KUBE_MASTER_URL} --ginkgo.skip=(should.allow.egress.access.to.server.in.CIDR.block)|(should.allow.ingress.access.from.updated.pod)"
```
This only test the named port tests, `.*on.one.named.port` should be removed.

jrajahalme replied:
removed.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -70,9 +70,9 @@ var _ = Describe("K8sPolicyTest", func() {
-		l7Policy = helpers.ManifestGet(kubectl.BasePath(), "l7-policy.yaml")
+		demoPath = helpers.ManifestGet(kubectl.BasePath(), "demo-named-port.yaml")
+		l3Policy = helpers.ManifestGet(kubectl.BasePath(), "l3-l4-policy-named-port.yaml")
+		l7Policy = helpers.ManifestGet(kubectl.BasePath(), "l7-policy-named-port.yaml")
```
This is now testing named ports but what about non-named ports?

jrajahalme replied:
Fixed this by duplicating some tests to be performed with named ports.

aanm commented on pkg/policy/l4.go:
```diff
@@ -804,11 +892,33 @@ func (l4 *L4Policy) insertUser(user *EndpointPolicy) {
+		// resolve named port
+		if port == 0 && l4Filter.PortName != "" {
+			var err error
+			port, err = epPolicy.NamedPortsMap.GetNamedPort(l4Filter.PortName, proto)
+			if err != nil {
```
Shouldn't we log the error?

jrajahalme replied:
It was already logged, but changed this to use `WithError(err)`.

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -311,14 +317,9 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
-			}
-
-			return p.Annotations[annotation.ProxyVisibility], nil
+			return visibilityAnnotation, nil
```
This change seems unrelated with the change. This seems to introduce a bug because we are not getting the most recent visibilityAnnotation as we were previously doing.

jrajahalme replied:
Did not realize that the annotations could change before the function is executed (can they?). Reverted.

aanm commented on daemon/cmd/state.go:
```diff
@@ -92,11 +92,18 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
-				_, _, annotations, err := d.fetchK8sLabelsAndAnnotations(ns, podName)
-				if err != nil {
-					return "", err
+			cp, _, _, annotations, err := d.fetchK8sLabelsAndAnnotations(ep.K8sNamespace, ep.K8sPodName)
```
This seems to introduce a regression as well since the annotations are not the ones up to dates when the `UpdateVisibilityPolicy` is executed.

jrajahalme replied:
Regression reverted.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -179,15 +179,25 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+					// skip redirects for which a proxyID cannot be created.
+					// Right now this happens only due to named port mapping not existing and
+					// we'll be called again when the mapping is available.
+					log.WithError(err).WithField(logfields.EndpointID, e.ID).Debug("Skipping adding redirect")
```
Shouldn't it Error instead of Debug?

jrajahalme replied:
There are two scenarios under which this can happen:
1) Ingress policy refers to a port name that is not defined for the POD. In that case the redirect will not be added, but also the L4 policy allow will not be added to the bpf policy map, so traffic to the named port with an unknown port number will not be allowed. This is correct if the named port is missing due to POD actually not listening on that port. If the port name missing from the POD spec is an error, then traffic will be dropped, and maybe a Warning log message would be helpful in figuring out what is going on. IMO Error is too much.

2) Egress policy refers to a port name that is not defined. Egress policies are defined in terms of the destination port number, and typically the named port would be specified in the POD spec of the destination POD. If it is, it may not be propagated to the source node yet. In this case the policy will be regenerated after the port name mapping arrives. If no POD in the cluster defines the port name, then the traffic to that named port will be dropped. If the missing port name is due to no POD actually listening on that port then the behavior is correct. If the port name is missing from the (destination) POD spec by accident, then a Warning log message may be helpful.

=> Will change from Debug to Warning :-)

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -1481,16 +1512,13 @@ func (e *Endpoint) RunMetadataResolver(resolveMetadata MetadataResolverCB) {
-					_, _, annotations, err := resolveMetadata(ns, podName)
-					if err != nil {
-						return "", err
-					}
```
Same regression as point it out in other occasion.

jrajahalme replied:
reverted.

aanm commented on pkg/endpoint/restore.go:
```diff
@@ -163,7 +163,7 @@ func (e *Endpoint) RegenerateAfterRestore() error {
 	if buildSuccess := <-e.Regenerate(regenerationMetadata); !buildSuccess {
 		scopedLog.Warn("Failed while regenerating endpoint")
-		return fmt.Errorf("failed while regenerating endpoint")
+		return fmt.Errorf("Failed while regenerating endpoint")
```
This does not seem related with the functionality introduced by this PR. It would be nice if code refactoring would be done in a different PR.

jrajahalme replied:
reverted

aanm commented on pkg/envoy/server.go:
```diff
@@ -935,27 +944,39 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool)
 		}
 
+		// Add rules to a new or existing entry for this port
+		key := fmt.Sprintf("%d/%s", port, l4.Protocol)
```
Might worth create a function that returns the a Key when given a port and a protocol.

jrajahalme replied:
This is actually used in quite a many places:
```
$ git grep "\"%d/%s\""
cilium/cmd/bpf_policy_get.go:                   port = fmt.Sprintf("%d/%s", dport, proto.String())
pkg/envoy/server.go:            key := fmt.Sprintf("%d/%s", port, l4.Protocol)
pkg/policy/l4.go:                       port := fmt.Sprintf("%d/%s", l4Ctx.Port, lwrProtocol)
pkg/policy/policy.go:           dports = append(dports, fmt.Sprintf("%d/%s", dport.Port, dport.Protocol))
pkg/policy/visibility.go:                       pp := fmt.Sprintf("%d/%s", portInt, prot.String())
```

But I will leave this for a future enhancement for now. The map in this case is a local one, so there need not be a contract about the map key with anyone else.

aanm commented on pkg/k8s/types/types.go:
```diff
@@ -49,8 +50,9 @@ type SlimCNP struct {
-	ContainerPort int32
+	Name          string `json:"name,omitempty"`
+	Protocol      string `json:"protocol,omitempty"`
+	ContainerPort int32  `json:"containerPort"`
```
why json tags?

jrajahalme replied:
Recall adding them due to empty name popping up in (unit) tests, but can't be sure, will test again without.

jrajahalme replied:
I can't find any reason why the json tags should be here, so taking them away :-)

aanm commented on pkg/policy/api/l4.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016-2019 Authors of Cilium
+// Copyright 2020 Authors of Cilium
```
Shouldn't this be 2016-2020?

jrajahalme replied:
Yes :-)

aanm commented on pkg/policy/l4.go:
```diff
@@ -180,7 +182,8 @@ const (
 	// is allowed at L4.
-	Port int `json:"port"`
+	Port     int    `json:"port"`
+	PortName string `json:"port-name"`
```
This could have `omitempty`, no?

jrajahalme replied:
Correct.

jrajahalme committed:
policy: Pass named k8s ContainerPorts to Endpoint

Store the set of named ContainerPorts in Endpoint. The ports are
immutable after Endpoint has been created.

Signed-off-by: Andre Martins <andre@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
api: Add named ports

Add named ports to endpoint status.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
